content/6_Advanced/SCC.mdx

```diff
@@ -183,6 +183,86 @@ public class Main {
 ```
 
 </JavaSection>
+<PySection>
+
+<Warning>
+
+Submit the solution in PyPy3 to avoid TLE.
+
+</Warning>
+
+```py
+order = []
+
+# calculates the order in which nodes are processed
+def dfs1(v: int):
+	stack = [v]
+	while stack:
+		node = stack[-1]
+		if not vis[node]:
+			vis[node] = True
+		finished = True
+		for u in adj[node]:
+			if not vis[u]:
+				stack.append(u)
+				finished = False
+				break
+		if finished:
+			order.append(stack.pop())
+		else:
+			vis[node] = True
+
+
+"""
+traverse one SCC at a time and
+assign component ID to all nodes in one SCC
+"""
+
+
+def dfs2(x: int, comp: int):
+	stack = [x]
+	while stack:
+		node = stack.pop()
+		if not vis[node]:
+			vis[node] = True
+			id[node] = comp
+			for u in adj_t[node]:
+				if not vis[u]:
+					stack.append(u)
+
+
+n, m = map(int, input().split())
+
+adj = [[] for _ in range(n)]
+adj_t = [[] for _ in range(n)]  # adj_t is the transpose of adj
```

Comment:
is this like the reversed adjacency list

Reply:
yes

Reply:
rename it to `rev_adj`, much more intuitive than "transpose"

Reply:
Done